File: app/src/main/java/com/example/data/remote/TokenStore.kt
package com.example.data.remote

import android.content.Context
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import android.util.Base64
import android.util.Log
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.booleanPreferencesKey
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import com.example.BuildConfig
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch
import java.security.KeyStore
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

// Preferences extension delegate
val Context.tokenDataStore: DataStore<Preferences> by preferencesDataStore(name = "curio_tokens")

/**
 * Persists and retrieves OAuth credentials using Keystore-backed AES-GCM encryption
 * over Jetpack Preferences DataStore.
 */
class TokenStore(private val context: Context) {

    companion object {
        private const val TAG = "TokenStore"
        private const val KEY_ALIAS = "CurioTokenStoreKey"
        private const val ANDROID_KEY_STORE = "AndroidKeyStore"
        private const val TRANSFORMATION = "AES/GCM/NoPadding"

        private val KEY_ACCESS_TOKEN_SURFACE = stringPreferencesKey("access_token_surface")
        private val KEY_REFRESH_TOKEN_SURFACE = stringPreferencesKey("refresh_token_surface")
        private val KEY_USER_ID = stringPreferencesKey("user_id")
        private val KEY_USERNAME = stringPreferencesKey("user_username")
        private val KEY_NAME = stringPreferencesKey("user_name")
        private val KEY_PROFILE_IMAGE_URL = stringPreferencesKey("user_profile_image_url")
        private val KEY_HF_TOKEN = stringPreferencesKey("hf_token_surface")
        private val KEY_XAI_KEY = stringPreferencesKey("xai_key_surface")
        private val KEY_X_CLIENT_ID = stringPreferencesKey("x_oauth_client_id")
        private val KEY_ALLOW_AGENT_WRITES = booleanPreferencesKey("allow_agent_writes")
        private val KEY_USE_DYNAMIC_COLOR = booleanPreferencesKey("use_dynamic_color")
        private val KEY_THEME_SETTING = stringPreferencesKey("theme_setting")
        private val KEY_GLASS_TIER_OVERRIDE = stringPreferencesKey("glass_tier_override")

        @Volatile private var debugFallbackKey: javax.crypto.SecretKey? = null

        /** Upper bound on a legitimate envelope IV length (AES-GCM uses 12; allow headroom). */
        private const val MAX_IV_BYTES = 64
    }

    init {
        getOrCreateSecretKey()
    }

    /**
     * Observable mapping of current logged in X userId
     */
    val userIdFlow: Flow<String?> = context.tokenDataStore.data.map { preferences ->
        preferences[KEY_USER_ID]
    }

    /**
     * Observable X handle (without the leading @) of the current logged in account.
     */
    val usernameFlow: Flow<String?> = context.tokenDataStore.data.map { preferences ->
        preferences[KEY_USERNAME]
    }

    /**
     * Observable X display name of the current logged in account.
     */
    val nameFlow: Flow<String?> = context.tokenDataStore.data.map { preferences ->
        preferences[KEY_NAME]
    }

    /**
     * Observable X profile photo URL of the current logged in account (the "_normal" 48×48
     * variant as returned by the API), or null if not fetched yet. Plain text: it's a public
     * CDN URL, not a credential.
     */
    val profileImageUrlFlow: Flow<String?> = context.tokenDataStore.data.map { preferences ->
        preferences[KEY_PROFILE_IMAGE_URL]
    }

    /** One-shot read of the stored profile photo URL, or null. */
    suspend fun getProfileImageUrl(): String? =
        context.tokenDataStore.data.first()[KEY_PROFILE_IMAGE_URL]?.takeIf { it.isNotBlank() }

    /** Persists the public account profile fields fetched from GET /2/users/me. */
    suspend fun saveProfile(username: String?, name: String?, profileImageUrl: String?) {
        context.tokenDataStore.edit { preferences ->
            if (!username.isNullOrBlank()) preferences[KEY_USERNAME] = username
            if (!name.isNullOrBlank()) preferences[KEY_NAME] = name
            if (!profileImageUrl.isNullOrBlank()) preferences[KEY_PROFILE_IMAGE_URL] = profileImageUrl
        }
    }

    /**
     * Whether on-device AI agents / system assistants may invoke the *write* AppFunctions
     * (add bookmark, add note, toggle favourite). Defaults to true (current behaviour); the user
     * can revoke it in Settings. Read-only discovery/detail functions are unaffected. This is the
     * only agent-scoping control available on appfunctions alpha09, which does not expose the
     * calling package to function code (see CurioFunctions).
     */
    val allowAgentWritesFlow: Flow<Boolean> = context.tokenDataStore.data.map { preferences ->
        preferences[KEY_ALLOW_AGENT_WRITES] ?: true
    }

    /** Whether Material You dynamic colors are enabled (Settings → Aesthetics). Defaults to false. */
    val useDynamicColorFlow: Flow<Boolean> = context.tokenDataStore.data.map { preferences ->
        preferences[KEY_USE_DYNAMIC_COLOR] ?: false
    }

    /** OS dark-style preference enum name; defaults to DARK when unset. */
    val themeSettingFlow: Flow<String?> = context.tokenDataStore.data.map { preferences ->
        preferences[KEY_THEME_SETTING]
    }

    /** Manual glass-tier override enum name, or null when set to Auto. */
    val glassTierOverrideFlow: Flow<String?> = context.tokenDataStore.data.map { preferences ->
        preferences[KEY_GLASS_TIER_OVERRIDE]
    }

    /** Returns the currently logged-in X userId, or null if not signed in. */
    suspend fun getUserId(): String? =
        context.tokenDataStore.data.first()[KEY_USER_ID]

    /**
     * Checks if we have cached credentials.
     */
    suspend fun hasTokens(): Boolean {
        val prefs = context.tokenDataStore.data.first()
        return !prefs[KEY_ACCESS_TOKEN_SURFACE].isNullOrEmpty()
    }

    /**
     * Resolves plain-text decrypted Access Token
     */
    suspend fun getAccessToken(): String? {
        val encryptedBase64 = context.tokenDataStore.data.first()[KEY_ACCESS_TOKEN_SURFACE] ?: return null
        return decrypt(encryptedBase64)
    }

    /**
     * Resolves plain-text decrypted Refresh Token
     */
    suspend fun getRefreshToken(): String? {
        val encryptedBase64 = context.tokenDataStore.data.first()[KEY_REFRESH_TOKEN_SURFACE] ?: return null
        return decrypt(encryptedBase64)
    }

    /**
     * Saves credentials safely using crypto envelopes.
     */
    suspend fun saveTokens(
        accessToken: String,
        refreshToken: String?,
        userId: String,
        username: String? = null,
        name: String? = null
    ) {
        val encryptedAccess = encrypt(accessToken)
        val encryptedRefresh = refreshToken?.let { encrypt(it) } ?: ""

        context.tokenDataStore.edit { preferences ->
            preferences[KEY_ACCESS_TOKEN_SURFACE] = encryptedAccess
            preferences[KEY_REFRESH_TOKEN_SURFACE] = encryptedRefresh
            preferences[KEY_USER_ID] = userId
            // Null means "not provided by this caller", NOT "remove": the background token
            // refresh saves rotated tokens without profile fields, and removing here made the
            // account name/handle vanish from the UI after the first ~2h refresh. Removal
            // happens only via [clear] at sign-out.
            if (username != null) preferences[KEY_USERNAME] = username
            if (name != null) preferences[KEY_NAME] = name
        }
    }

    /**
     * Resolves the plain-text decrypted Hugging Face token used to fetch gated model weights, or
     * null if none was saved. Independent of the X session — intentionally not removed by [clear].
     */
    suspend fun getHuggingFaceToken(): String? {
        val encryptedBase64 = context.tokenDataStore.data.first()[KEY_HF_TOKEN] ?: return null
        return decrypt(encryptedBase64)
    }

    /** Persists the Hugging Face token under Keystore-backed AES-GCM encryption. */
    suspend fun saveHuggingFaceToken(token: String) {
        val encrypted = encrypt(token)
        context.tokenDataStore.edit { preferences ->
            preferences[KEY_HF_TOKEN] = encrypted
        }
    }

    /**
     * Resolves the user-supplied xAI API key (encrypted), or null if none was saved. Independent
     * of the X session — intentionally not removed by [clear].
     */
    suspend fun getXaiKey(): String? {
        val encryptedBase64 = context.tokenDataStore.data.first()[KEY_XAI_KEY] ?: return null
        return decrypt(encryptedBase64)
    }

    /** Persists (or clears, with a blank value) the xAI API key under AES-GCM encryption. */
    suspend fun saveXaiKey(key: String) {
        context.tokenDataStore.edit { preferences ->
            if (key.isBlank()) preferences.remove(KEY_XAI_KEY)
            else preferences[KEY_XAI_KEY] = encrypt(key)
        }
    }

    /**
     * User-supplied X OAuth client ID (BYOK for bookmark sync), or null if none was saved.
     * Stored in plain text intentionally: an OAuth public-client ID is not a secret (it appears
     * verbatim in the browser authorize URL), and skipping the Keystore round-trip means a
     * Keystore hiccup can never silently knock sync back to the built-in client.
     * Independent of the X session — intentionally not removed by [clear], so signing out to
     * re-login (required after changing the ID) doesn't wipe it.
     */
    suspend fun getXClientId(): String? =
        context.tokenDataStore.data.first()[KEY_X_CLIENT_ID]?.takeIf { it.isNotBlank() }

    /** Persists (or clears, with a blank value) the user-supplied X OAuth client ID. */
    suspend fun saveXClientId(clientId: String) {
        context.tokenDataStore.edit { preferences ->
            if (clientId.isBlank()) preferences.remove(KEY_X_CLIENT_ID)
            else preferences[KEY_X_CLIENT_ID] = clientId.trim()
        }
    }

    /**
     * The single choke point for choosing the X OAuth client ID: user-supplied (BYOK) first, then
     * the secrets-plugin CLIENT_ID, then the gradle-injected X_CLIENT_ID default. Login AND token
     * refresh must both resolve through here — they previously preferred the two BuildConfig
     * fields in OPPOSITE orders, so when the fields differed the refresh went out with a client_id
     * that didn't match the one that minted the token, and every sync after the ~2h access-token
     * expiry failed with HTTP 401.
     */
    suspend fun resolveXClientId(): String {
        getXClientId()?.let { return it }
        return BuildConfig.X_CLIENT_ID
    }

    /** One-shot read of the agent-write permission, for the AppFunction write gate. */
    suspend fun isAgentWritesAllowed(): Boolean =
        context.tokenDataStore.data.first()[KEY_ALLOW_AGENT_WRITES] ?: true

    /** Persists the agent-write permission toggled from Settings. */
    suspend fun setAgentWritesAllowed(allowed: Boolean) {
        context.tokenDataStore.edit { preferences ->
            preferences[KEY_ALLOW_AGENT_WRITES] = allowed
        }
    }

    /** Persists the Material You dynamic-color toggle from Settings. */
    suspend fun setUseDynamicColor(enabled: Boolean) {
        context.tokenDataStore.edit { preferences ->
            preferences[KEY_USE_DYNAMIC_COLOR] = enabled
        }
    }

    /** Persists the OS dark-style preference from Settings. */
    suspend fun setThemeSetting(setting: String) {
        context.tokenDataStore.edit { preferences ->
            preferences[KEY_THEME_SETTING] = setting
        }
    }

    /** Persists (or clears) the manual glass-tier override from Settings. */
    suspend fun setGlassTierOverride(tier: String?) {
        context.tokenDataStore.edit { preferences ->
            if (tier.isNullOrBlank()) preferences.remove(KEY_GLASS_TIER_OVERRIDE)
            else preferences[KEY_GLASS_TIER_OVERRIDE] = tier
        }
    }

    /**
     * Purges all locally cached session elements completely.
     */
    suspend fun clear() {
        context.tokenDataStore.edit { preferences ->
            preferences.remove(KEY_ACCESS_TOKEN_SURFACE)
            preferences.remove(KEY_REFRESH_TOKEN_SURFACE)
            preferences.remove(KEY_USER_ID)
            preferences.remove(KEY_USERNAME)
            preferences.remove(KEY_NAME)
            preferences.remove(KEY_PROFILE_IMAGE_URL)
        }
    }

    // --- CRYPTO LOGIC ---

    /**
     * Synchronous credential purge used by the KeyStore failure path in release builds.
     * Runs on whatever thread calls [getOrCreateSecretKey], which is fine — DataStore
     * write operations are serialized internally.
     */
    private fun clearAll() {
        kotlinx.coroutines.CoroutineScope(kotlinx.coroutines.Dispatchers.IO).launch {
            try { clear() } catch (e: Exception) { /* best-effort */ }
        }
    }

    private fun getOrCreateSecretKey(): SecretKey {
        try {
            val keyStore = KeyStore.getInstance(ANDROID_KEY_STORE).apply { load(null) }
            keyStore.getKey(KEY_ALIAS, null)?.let { return it as SecretKey }

            val keyGenerator = KeyGenerator.getInstance(
                KeyProperties.KEY_ALGORITHM_AES,
                ANDROID_KEY_STORE
            )

            val spec = KeyGenParameterSpec.Builder(
                KEY_ALIAS,
                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT
            )
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                .build()

            keyGenerator.init(spec)
            return keyGenerator.generateKey()
        } catch (e: Exception) {
            // Software fallback only in debug builds (JVM unit / screenshot tests where
            // AndroidKeyStore provider is absent). In release builds we clear credentials and
            // surface an exception so the caller can handle the failure explicitly.
            if (BuildConfig.DEBUG) {
                return debugFallbackKey ?: synchronized(this) {
                    debugFallbackKey ?: javax.crypto.spec.SecretKeySpec(
                        ByteArray(16) { 0x55.toByte() }, // 128-bit stable fallback test key
                        "AES"
                    ).also { debugFallbackKey = it }
                }
            } else {
                clearAll()
                throw SecurityException("AndroidKeyStore unavailable; credentials cleared", e)
            }
        }
    }

    private fun encrypt(plainText: String): String {
        val key = getOrCreateSecretKey()
        val cipher = Cipher.getInstance(TRANSFORMATION).apply {
            init(Cipher.ENCRYPT_MODE, key)
        }
        val cipherText = cipher.doFinal(plainText.toByteArray(Charsets.UTF_8))
        val iv = cipher.iv

        // Combine IV and CipherText: [IV_Length (4 bytes)][IV][CipherText]
        val combined = ByteArray(4 + iv.size + cipherText.size)
        combined[0] = (iv.size shr 24).toByte()
        combined[1] = (iv.size shr 16).toByte()
        combined[2] = (iv.size shr 8).toByte()
        combined[3] = iv.size.toByte()

        System.arraycopy(iv, 0, combined, 4, iv.size)
        System.arraycopy(cipherText, 0, combined, 4 + iv.size, cipherText.size)

        return Base64.encodeToString(combined, Base64.DEFAULT)
    }

    private fun decrypt(encryptedBase64: String): String? {
        return try {
            val combined = Base64.decode(encryptedBase64, Base64.DEFAULT)
            if (combined.size < 4) return null

            val ivSize = ((combined[0].toInt() and 0xFF) shl 24) or
                    ((combined[1].toInt() and 0xFF) shl 16) or
                    ((combined[2].toInt() and 0xFF) shl 8) or
                    (combined[3].toInt() and 0xFF)

            // Harden the envelope header against corruption/hostility before allocating:
            // a huge or negative length used to slip past the size check via Int overflow
            // (4 + ivSize wraps negative), then ByteArray(huge) threw OutOfMemoryError —
            // an Error the catch below cannot intercept. Real GCM IVs are 12 bytes.
            if (ivSize <= 0 || ivSize > MAX_IV_BYTES || combined.size < 4 + ivSize) return null

            val iv = ByteArray(ivSize)
            System.arraycopy(combined, 4, iv, 0, ivSize)

            val cipherTextSize = combined.size - 4 - ivSize
            val cipherText = ByteArray(cipherTextSize)
            System.arraycopy(combined, 4 + ivSize, cipherText, 0, cipherTextSize)

            val key = getOrCreateSecretKey()
            val cipher = Cipher.getInstance(TRANSFORMATION).apply {
                init(Cipher.DECRYPT_MODE, key, GCMParameterSpec(128, iv))
            }
            val decryptedBytes = cipher.doFinal(cipherText)
            String(decryptedBytes, Charsets.UTF_8)
        } catch (e: Exception) {
            Log.w(TAG, "Decryption/KeyStore error", e)
            null
        }
    }
}

File: ios/Curio/Curio/Auth/TokenStore.swift
import Foundation
import Combine
import Security
import os

/// Persists and retrieves OAuth credentials (and the HF / xAI secrets) in the iOS **Keychain**.
/// Ports `data/remote/TokenStore.kt`.
///
/// Android used AndroidKeyStore AES-GCM over a Preferences DataStore named `curio_tokens`. On iOS
/// the Keychain already provides hardware-backed, at-rest-encrypted secure storage, so we take
/// **Path A** (direct Keychain, CONVENTIONS §"AndroidKeyStore AES-GCM + DataStore" / §9) — each
/// secret is a `kSecClassGenericPassword` item with
/// `kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly` (available after first unlock, never
/// migrated to a new device — mirroring the device-bound, non-exportable AndroidKeyStore key). The
/// per-item AES-GCM envelope the Android code hand-rolled is unnecessary and intentionally dropped:
/// the Keychain is the encryption boundary.
///
/// **Account-string fidelity (CONVENTIONS §"Persistence-key stability is sacred"):** the Keychain
/// `kSecAttrAccount` values are the *exact* DataStore preference keys from Android
/// (`access_token_surface`, `refresh_token_surface`, `user_id`, `user_username`, `user_name`,
/// `hf_token_surface`, `xai_key_surface`) so a migrated install (or a shared semantic contract)
/// reads/writes the same logical slots. Never rename these.
///
/// **Semantics preserved verbatim from Kotlin:**
/// - `hasTokens()` = access-token item present AND non-empty (`!isNullOrEmpty`).
/// - `saveTokens(refreshToken: nil)` writes an empty string for the refresh slot (Kotlin
///   `refreshToken?.let { encrypt(it) } ?: ""`), so a later `getRefreshToken()` returns `nil`
///   (empty → treated as absent, see `getRefreshToken`).
/// - `saveTokens(username: nil)` / `name: nil` **removes** that item (Kotlin `preferences.remove`).
/// - `saveXaiKey("")` / blank **deletes** the xAI item (Kotlin `if (key.isBlank()) remove`).
/// - `clear()` removes ONLY the X-session items (access/refresh/userId/username/name); the HF token
///   and xAI key intentionally survive (Kotlin `clear` leaves `KEY_HF_TOKEN` / `KEY_XAI_KEY`).
///
/// **Reactive identity (CONVENTIONS §11):** Android exposed `userIdFlow` / `usernameFlow` /
/// `nameFlow` as hot DataStore-backed `Flow<String?>`. Here three `CurrentValueSubject<String?,
/// Never>` carry the current value and re-emit on every `saveTokens` / `clear`, exposed as
/// `AnyPublisher`s. They are seeded from the Keychain at construction. The subjects are reference
/// types, created before isolation is established, and are `nonisolated` so `AuthRepositoryImpl`'s
/// boot restore and the auth controller can subscribe without hopping onto the actor.
///
/// `actor` per CONVENTIONS §5 (owns mutable secure-storage access; serializes reads/writes).
actor TokenStore {

    // MARK: - Account keys (mirror Android DataStore preference keys EXACTLY)

    private enum Account {
        static let accessToken = "access_token_surface"
        static let refreshToken = "refresh_token_surface"
        static let userId = "user_id"
        static let username = "user_username"
        static let name = "user_name"
        static let huggingFaceToken = "hf_token_surface"
        static let xaiKey = "xai_key_surface"
    }

    /// Keychain service namespace for all Curio secrets (groups the items; mirrors the
    /// `curio_tokens` DataStore name).
    private static let service = "com.curio.app.tokens"

    /// UserDefaults key for the agent-write permission (Android DataStore `allow_agent_writes`).
    /// A plain preference, not a secret — it lives in UserDefaults, not the Keychain.
    private static let allowAgentWritesKey = "allow_agent_writes"

    private static let logger = Logger(subsystem: "com.curio.app", category: "TokenStore")

    // MARK: - Reactive identity subjects (nonisolated, reference-typed, thread-safe)

    // `nonisolated(unsafe)`: Combine subjects are reference-typed and thread-safe for send/subscribe
    // but not formally `Sendable`; we vouch for the cross-isolation access.
    private nonisolated(unsafe) let userIdSubject: CurrentValueSubject<String?, Never>
    private nonisolated(unsafe) let usernameSubject: CurrentValueSubject<String?, Never>
    private nonisolated(unsafe) let nameSubject: CurrentValueSubject<String?, Never>
    private nonisolated(unsafe) let allowAgentWritesSubject: CurrentValueSubject<Bool, Never>

    /// Observable mapping of the current logged-in X userId (Android `userIdFlow`).
    nonisolated var userIdPublisher: AnyPublisher<String?, Never> {
        userIdSubject.eraseToAnyPublisher()
    }

    /// Observable X handle (without the leading `@`) of the current account (Android `usernameFlow`).
    nonisolated var usernamePublisher: AnyPublisher<String?, Never> {
        usernameSubject.eraseToAnyPublisher()
    }

    /// Observable X display name of the current account (Android `nameFlow`).
    nonisolated var namePublisher: AnyPublisher<String?, Never> {
        nameSubject.eraseToAnyPublisher()
    }

    /// Whether on-device AI agents / system assistants may invoke the *write* App Intents
    /// (add bookmark, add note, toggle favourite). Defaults to true; the user can revoke it in
    /// Settings. Read-only discovery/detail intents are unaffected. Android `allowAgentWritesFlow`.
    nonisolated var allowAgentWritesPublisher: AnyPublisher<Bool, Never> {
        allowAgentWritesSubject.eraseToAnyPublisher()
    }

    // MARK: - Init

    init() {
        // Seed the subjects from the Keychain so subscribers see the persisted identity immediately
        // (mirrors DataStore replaying the last value to a new collector). Reads here are static so
        // they do not touch actor-isolated state.
        userIdSubject = CurrentValueSubject(Self.readString(account: Account.userId))
        usernameSubject = CurrentValueSubject(Self.readString(account: Account.username))
        nameSubject = CurrentValueSubject(Self.readString(account: Account.name))
        allowAgentWritesSubject = CurrentValueSubject(Self.readAllowAgentWrites())
    }

    // MARK: - Session reads

    /// Returns the currently logged-in X userId, or `nil` if not signed in. Plain `suspend` →
    /// `async` (no throws), per CONVENTIONS §3.
    func getUserId() -> String? {
        Self.readString(account: Account.userId)
    }

    /// Checks whether cached credentials exist: the access-token item must be present AND non-empty
    /// (Kotlin `!prefs[KEY_ACCESS_TOKEN_SURFACE].isNullOrEmpty()`).
    func hasTokens() -> Bool {
        guard let token = Self.readString(account: Account.accessToken) else { return false }
        return !token.isEmpty
    }

    /// Resolves the plain-text access token, or `nil` if none was saved.
    func getAccessToken() -> String? {
        Self.readString(account: Account.accessToken)
    }

    /// Resolves the stored X handle (without the leading `@`), or `nil`. Mirrors reading the
    /// Android `usernameFlow`'s current value (used by the boot restore).
    func getUsername() -> String? {
        Self.readString(account: Account.username)
    }

    /// Resolves the stored X display name, or `nil`. Mirrors reading the Android `nameFlow`'s
    /// current value (used by the boot restore).
    func getName() -> String? {
        Self.readString(account: Account.name)
    }

    /// Resolves the plain-text refresh token, or `nil` if none was saved. The empty-string sentinel
    /// written for a missing refresh token (see `saveTokens`) is normalised back to `nil`.
    func getRefreshToken() -> String? {
        guard let token = Self.readString(account: Account.refreshToken), !token.isEmpty else {
            return nil
        }
        return token
    }

    // MARK: - Session writes

    /// Saves credentials. Mirrors the Kotlin write transaction exactly:
    /// - refresh token: persisted, or empty string when `nil`;
    /// - username / name: written when non-nil, otherwise the item is **removed**.
    func saveTokens(
        accessToken: String,
        refreshToken: String?,
        userId: String,
        username: String? = nil,
        name: String? = nil
    ) {
        Self.write(account: Account.accessToken, value: accessToken)
        Self.write(account: Account.refreshToken, value: refreshToken ?? "")
        Self.write(account: Account.userId, value: userId)

        if let username {
            Self.write(account: Account.username, value: username)
        } else {
            Self.delete(account: Account.username)
        }

        if let name {
            Self.write(account: Account.name, value: name)
        } else {
            Self.delete(account: Account.name)
        }

        // Re-emit identity (mirrors DataStore Flow re-emission on edit).
        userIdSubject.send(userId)
        usernameSubject.send(username)
        nameSubject.send(name)
    }

    // MARK: - Hugging Face token (survives clear())

    /// Resolves the Hugging Face token used to fetch gated model weights, or `nil`. Independent of
    /// the X session — intentionally **not** removed by `clear()`.
    func getHuggingFaceToken() -> String? {
        Self.readString(account: Account.huggingFaceToken)
    }

    /// Persists the Hugging Face token.
    func saveHuggingFaceToken(_ token: String) {
        Self.write(account: Account.huggingFaceToken, value: token)
    }

    // MARK: - xAI key (survives clear(); blank deletes)

    /// Resolves the user-supplied xAI API key, or `nil`. Independent of the X session — intentionally
    /// **not** removed by `clear()`.
    func getXaiKey() -> String? {
        Self.readString(account: Account.xaiKey)
    }

    /// Persists (or clears, with a blank value) the xAI API key. Mirrors Kotlin: a blank key
    /// **removes** the item rather than storing whitespace.
    func saveXaiKey(_ key: String) {
        if key.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            Self.delete(account: Account.xaiKey)
        } else {
            Self.write(account: Account.xaiKey, value: key)
        }
    }

    // MARK: - Agent write permission (plain preference; survives clear())

    /// One-shot read of the agent-write permission, for the App Intent write gate
    /// (Android `isAgentWritesAllowed`). Defaults to true when never set.
    func isAgentWritesAllowed() -> Bool {
        Self.readAllowAgentWrites()
    }

    /// Persists the agent-write permission toggled from Settings (Android `setAgentWritesAllowed`).
    func setAgentWritesAllowed(_ allowed: Bool) {
        UserDefaults.standard.set(allowed, forKey: Self.allowAgentWritesKey)
        allowAgentWritesSubject.send(allowed)
    }

    /// Reads the persisted permission, defaulting to true when the key was never written
    /// (mirrors Android `preferences[KEY_ALLOW_AGENT_WRITES] ?: true`).
    private static func readAllowAgentWrites() -> Bool {
        UserDefaults.standard.object(forKey: allowAgentWritesKey) as? Bool ?? true
    }

    // MARK: - Session purge

    /// Purges the X-session items ONLY (access/refresh/userId/username/name). The HF token and xAI
    /// key intentionally survive (mirrors Kotlin `clear`, which leaves `KEY_HF_TOKEN` / `KEY_XAI_KEY`
    /// in place).
    func clear() {
        Self.delete(account: Account.accessToken)
        Self.delete(account: Account.refreshToken)
        Self.delete(account: Account.userId)
        Self.delete(account: Account.username)
        Self.delete(account: Account.name)

        userIdSubject.send(nil)
        usernameSubject.send(nil)
        nameSubject.send(nil)
    }

    // MARK: - Keychain primitives (static — used during init before isolation is established)

    /// Base query identifying a single generic-password item by service + account.
    private static func baseQuery(account: String) -> [String: Any] {
        [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account
        ]
    }

    /// Reads a UTF-8 string for `account`, or `nil` if absent / unreadable. Resilient (logs +
    /// returns `nil`) — never throws, mirroring the Android decrypt `catch → null` contract
    /// (CONVENTIONS §3 "Resilience contract").
    private static func readString(account: String) -> String? {
        var query = baseQuery(account: account)
        query[kSecReturnData as String] = true
        query[kSecMatchLimit as String] = kSecMatchLimitOne

        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        guard status == errSecSuccess else {
            if status != errSecItemNotFound {
                logger.warning("Keychain read failed (status: \(status))")
            }
            return nil
        }
        guard let data = item as? Data else { return nil }
        return String(data: data, encoding: .utf8)
    }

    /// Upserts a UTF-8 string for `account`. Performs an update-then-add so the item is replaced
    /// in place (mirrors DataStore `edit { … = value }` overwrite semantics).
    private static func write(account: String, value: String) {
        let data = Data(value.utf8)
        let query = baseQuery(account: account)

        let attributes: [String: Any] = [
            kSecValueData as String: data,
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        ]

        let updateStatus = SecItemUpdate(query as CFDictionary, attributes as CFDictionary)
        if updateStatus == errSecSuccess { return }

        if updateStatus == errSecItemNotFound {
            var addQuery = query
            addQuery[kSecValueData as String] = data
            addQuery[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
            let addStatus = SecItemAdd(addQuery as CFDictionary, nil)
            if addStatus != errSecSuccess {
                logger.warning("Keychain add failed (status: \(addStatus))")
            }
        } else {
            logger.warning("Keychain update failed (status: \(updateStatus))")
        }
    }

    /// Removes the item for `account` (idempotent — `errSecItemNotFound` is success).
    private static func delete(account: String) {
        let status = SecItemDelete(baseQuery(account: account) as CFDictionary)
        if status != errSecSuccess && status != errSecItemNotFound {
            logger.warning("Keychain delete failed (status: \(status))")
        }
    }
}
